// element-plus/packages/utils/lib/props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.componentSize = exports.keyOf = exports.buildProp = void 0;
/**
 * @description Build prop. It can better optimize prop types
 * @description 生成 prop，能更好地优化类型
 * @example
  // limited options
  // the type will be PropType<'light' | 'dark'>
  buildProp({
    type: String,
    values: ['light', 'dark'],
  } as const)
  * @example
  // limited options and other types
  // the type will be PropType<'small' | 'medium' | number>
  buildProp({
    type: [String, Number],
    values: ['small', 'medium'],
    validator: (val: unknown): val is number => typeof val === 'number',
  } as const)
  @link for more explanation: https://github.com/element-plus/element-plus/pull/3341
 */
function buildProp({ values, required, default: defaultValue, type, validator, } = {}) {
    return {
        type: type,
        required: !!required,
        default: defaultValue,
        validator: values || validator
            ? (val) => {
                let vaild = false;
                if (values)
                    vaild || (vaild = [...values, defaultValue].includes(val));
                if (validator)
                    vaild || (vaild = validator(val));
                return vaild;
            }
            : undefined,
    };
}
exports.buildProp = buildProp;
function keyOf(arr) {
    return Object.keys(arr);
}
exports.keyOf = keyOf;
exports.componentSize = ['large', 'medium', 'small', 'mini'];
